### PR TITLE
[RFR] Make part of react-admin independant of mui

### PR DESCRIPTION
**Pull request by djhi**
- [x] Move pure logic parts of `react-admin` to `react-admin-core`
- [x] Rename `react-admin-core` to `ra-core`
- [x] Rename the current `react-admin` to `ra-ui-materialui`
- [x] Setup a new `react-admin` package as a *distribution* of `react-admin` with `ra-ui-materialui`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,7 +23,11 @@ run-graphql-demo: build ## run the demo example
 run-graphcool-demo: build ## run the demo example
 	@cd examples/graphcool-demo && yarn start
 
-build-react-admin:
+build-react-admin-core:
+	@rm -rf ./packages/react-admin-core/lib
+	@NODE_ENV=production ./node_modules/.bin/babel ./packages/react-admin-core/src -d ./packages/react-admin-core/lib --ignore spec.js,test.js
+
+build-react-admin: build-react-admin-core
```
Comment (fzaninotto) [inline Makefile]: I'd prefer that you add the `build-react-admin-core` as a predecessor to the `build` target, like all the other packages.

File: packages/react-admin-core/src/Admin.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
+import React, { createElement } from 'react';
```
Comment (fzaninotto) [inline packages/react-admin-core/src/Admin.js]: git considers this file as a creation, while you mostly moved it. This will mess up the history for blames.

File: packages/react-admin-core/src/AdminRouter.js [hunk 3]
```diff
@@ -96,18 +96,7 @@ export class AdminRouter extends Component {
             typeof children === 'function' ? childrenFromState : children;
 
         if (!childrenToRender || childrenToRender.length === 0) {
-            return (
```
Comment (fzaninotto) [inline packages/react-admin-core/src/AdminRouter.js]: this is a BC break, and your PR shouldn't break BC

File: packages/react-admin-core/src/reducer/admin/resource/index.js [hunk 4]
```diff
@@ -69,4 +69,6 @@ export default (
 export const getResources = state =>
     Object.keys(state).map(key => state[key].props);
 
-export const getReferenceResource = (state, props) => state[props.reference];
+export const getReferenceResource = (state, props) => {
```
Comment (fzaninotto) [inline packages/react-admin-core/src/reducer/admin/resource/index.js]: useless change

File: packages/react-admin-core/src/Admin.js (hunk 2 above)
Comment (djhi) [reply packages/react-admin-core/src/Admin.js]: Any solution ?

File: packages/react-admin-core/src/AdminRouter.js (hunk 3 above)
Comment (djhi) [reply packages/react-admin-core/src/AdminRouter.js]: This is a BC break for custom apps only. Solution would be to have this renamed as `CoreAdminRouter` and have an `AdminRouter` setting the default `loading` inside `ra-ui-materialui`. It won't be used by default though.

File: packages/ra-core/package.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
+{
+    "name": "ra-core",
+    "version": "2.0.0-beta2",
+    "description": "Core components of react-admin, a frontend Framework for building admin applications on top of REST services, using ES6, React",
+    "files": [
+        "*.md",
+        "docs/*.md",
```
Comment (fzaninotto) [inline packages/ra-core/package.json]: no docs in ra-core (for now)

File: packages/ra-core/package.json [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        "Gildas Garcia"
+    ],
+    "repository": "marmelab/react-admin",
+    "homepage": "https://github.com/marmelab/react-admin#readme",
+    "bugs": "https://github.com/marmelab/react-admin/issues",
+    "license": "MIT",
+    "devDependencies": {
+        "enzyme": "~3.1.0",
+        "enzyme-adapter-react-16": "~1.0.1",
+        "file-api": "~0.10.4",
```
Comment (fzaninotto) [inline packages/ra-core/package.json]: for what?

File: packages/ra-core/package.json [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        "lodash.debounce": "~4.0.8",
+        "lodash.defaultsdeep": "~4.6.0",
+        "lodash.get": "~4.4.2",
+        "lodash.pickby": "~4.6.0",
+        "lodash.set": "~4.3.2",
+        "lodash.uniq": "~4.5.0",
+        "node-polyglot": "2.2.2",
+        "prop-types": "~15.5.7",
+        "query-string": "~5.0.0",
+        "ra-language-english": "^2.0.0-beta2",
```
Comment (fzaninotto) [inline packages/ra-core/package.json]: should not be used here, only in `react-admin`

File: packages/ra-core/src/ui/delete/Delete.js [hunk 8]
```diff
@@ -174,72 +120,25 @@ export class Delete extends Component {
         const defaultTitle = translate('ra.page.delete', {
             name: `${resourceName}`,
             id,
-            data,
+            record,
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/delete/Delete.js]: bc break

File: packages/ra-core/src/ui/detail/Create.js [hunk 9]
```diff
@@ -78,8 +49,8 @@ const sanitizeRestProps = ({
  */
 class Create extends Component {
     getBasePath() {
-        const { location } = this.props;
-        return location.pathname
+        const { match } = this.props;
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/detail/Create.js]: why this change?

File: packages/ra-core/src/ui/delete/Delete.js [hunk 10]
```diff
@@ -174,72 +120,25 @@ export class Delete extends Component {
...
-            </div>
-        );
+        return children({
+            basePath,
+            defaultTitle,
+            handleSubmit: this.handleSubmit,
+            isLoading,
+            record,
+            goBack: this.goBack,
+        });
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/delete/Delete.js]: so you're no longer passing rest props to children, not good

File: packages/ra-core/src/reducer/admin/notifications.js [hunk 11]
```diff
@@ -18,4 +18,4 @@ export default (previousState = [], { type, payload }) => {
  * Returns the first available notification to show
  * @param {Object} state - Redux state 
  */
-export const getNotification = state => state.admin.notifications[0];
+export const getNotification = state => state[0];
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/admin/notifications.js]: please don't refactor that in your PR, it adds too much noise.

File: packages/ra-core/src/reducer/index.js [hunk 12]
```diff
@@ -25,4 +25,5 @@ export const getReferenceResource = (state, props) =>
     adminGetReferenceResource(state.admin, props);
 export const isLoggedIn = state => adminIsLoggedIn(state.admin);
 export const getLocale = state => adminGetLocale(state.i18n);
+export const getNotification = state => adminGetNotification(state.admin);
```
Comment (fzaninotto) [inline packages/ra-core/src/reducer/index.js]: please extract to another PR

File: packages/ra-core/src/ui/detail/Edit.js [hunk 13]
```diff
@@ -102,8 +68,8 @@ export class Edit extends Component {
     }
 
     getBasePath() {
-        const { location } = this.props;
-        return location.pathname
+        const { match } = this.props;
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/detail/Edit.js]: same, this change adds noise, please extract to another PR

File: packages/ra-core/src/ui/detail/Show.js [hunk 14]
```diff
@@ -134,52 +99,32 @@ export class Show extends Component {
         const defaultTitle = translate('ra.page.show', {
             name: `${resourceName}`,
             id,
-            data,
+            record,
+        });
+        return children({
+            actionProps: {
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/detail/Show.js]: will recreate a new object at each render, and therefore miss the identity check. You should inline them all.

File: packages/ra-core/src/ui/field/ReferenceField.js [hunk 15]
```diff
@@ -70,62 +59,30 @@ export class ReferenceField extends Component {
...
             children,
-            className,
-            classes,
             linkType,
             record,
             reference,
             referenceRecord,
             resource,
             source,
-            translateChoice,
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/field/ReferenceField.js]: why did you move that to the mui component? It should stay in the core IMO

File: packages/ra-core/src/ui/field/ReferenceManyField.js [hunk 16]
```diff
@@ -127,38 +119,26 @@ export class ReferenceManyField extends Component {
             basePath,
             isLoading,
         } = this.props;
-        if (React.Children.count(children) !== 1) {
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/field/ReferenceManyField.js]: this is the core component's responsibility

File: packages/ra-core/src/ui/field/ReferenceManyField.js [hunk 17]
```diff
@@ -198,13 +178,7 @@ function mapStateToProps(state, props) {
 const ComposedReferenceManyField = compose(
     connect(mapStateToProps, {
         crudGetManyReference: crudGetManyReferenceAction,
-    }),
-    withStyles(styles)
+    })
 )(ReferenceManyField);
 
-ComposedReferenceManyField.defaultProps = {
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/field/ReferenceManyField.js]: same, this is the core component's responsibility

File: packages/ra-core/src/ui/form/FormField.js [hunk 18]
```diff
@@ -3,7 +3,7 @@ import { connect } from 'react-redux';
 import PropTypes from 'prop-types';
 import { Field } from 'redux-form';
 
-import { initializeForm } from '../../actions';
+import { initializeForm } from 'ra-core';
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/form/FormField.js]: not good, I'm not fan of packages requiring themselves

File: packages/ra-core/src/ui/input/ReferenceArrayInput.js [hunk 19]
```diff
@@ -206,100 +177,49 @@ export class ReferenceArrayInput extends Component {
...
     crudGetMatching: PropTypes.func.isRequired,
     crudGetMany: PropTypes.func.isRequired,
     filter: PropTypes.object,
     filterToQuery: PropTypes.func.isRequired,
     input: PropTypes.object.isRequired,
     label: PropTypes.string,
-    matchingReferences: PropTypes.array,
+    matchingReferences: PropTypes.oneOfType([
+        PropTypes.array,
+        PropTypes.object,
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/input/ReferenceArrayInput.js]: in which case is it an object? Why did you change it in this PR?

File: packages/ra-core/src/ui/input/ReferenceInput.js [hunk 20]
```diff
@@ -215,101 +172,49 @@ export class ReferenceInput extends Component {
...
-        }
-
         const dataStatus = getDataStatus({
             input,
             matchingReferences,
             referenceRecord,
             translate,
         });
 
-        const translatedLabel = translate(
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/input/ReferenceInput.js]: shouldn't translation still be done in the core?

File: packages/ra-core/src/ui/input/ReferenceInput.js [hunk 21]
```diff
@@ -215,101 +172,49 @@ export class ReferenceInput extends Component {
...
-            basePath,
             onChange,
             children,
-            meta,
-            options,
             translate,
-            ...rest
         } = this.props;
 
-        if (React.Children.count(children) !== 1) {
```
Comment (fzaninotto) [inline packages/ra-core/src/ui/input/ReferenceInput.js]: core component responsibility IMO

File: packages/ra-ui-materialui/package.json [hunk 22]
```diff
@@ -0,0 +1,54 @@
+{
+    "name": "ra-ui-materialui",
+    "version": "2.0.0-beta2",
+    "description": "UI Components for react-admin with MaterialUI",
+    "files": [
+        "*.md",
+        "docs/*.md",
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/package.json]: no docs in this package

File: packages/ra-ui-materialui/package.json [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+        "Gildas Garcia"
+    ],
+    "repository": "marmelab/react-admin",
+    "homepage": "https://github.com/marmelab/react-admin#readme",
+    "bugs": "https://github.com/marmelab/react-admin/issues",
+    "license": "MIT",
+    "devDependencies": {
+        "enzyme": "~3.1.0",
+        "enzyme-adapter-react-16": "~1.0.1",
+        "file-api": "~0.10.4",
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/package.json]: used for what?

File: packages/ra-ui-materialui/package.json [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+        "full-icu": "~1.1.3",
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/package.json]: used for what?

File: packages/ra-ui-materialui/package.json [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+        "classnames": "~2.2.5",
+        "inflection": "~1.12.0",
+        "lodash.debounce": "~4.0.8",
+        "lodash.get": "~4.4.2",
+        "lodash.isequal": "~4.5.0",
+        "lodash.set": "~4.3.2",
+        "material-ui": "1.0.0-beta.32",
+        "material-ui-icons": "~1.0.0-beta.17",
+        "prop-types": "~15.5.7",
+        "ra-core": "^2.0.0-beta2",
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/package.json]: please use tilde

File: packages/ra-ui-materialui/src/detail/EditActions.js [hunk 26]
```diff
@@ -46,24 +46,24 @@ const sanitizeRestProps = ({
 const EditActions = ({
     basePath,
     className,
-    data,
+    record,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/EditActions.js]: bc break

File: packages/ra-ui-materialui/src/detail/Show.js [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+        >
+            {({
+                basePath,
+                defaultTitle,
+                isLoading,
+                record,
+                title,
+                version,
+            }) => {
+                const titleElement = record ? (
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/Show.js]: this one is repeated often enough to be extracted to a Component used by all views. It would allow you to reduce the indentation of this file.

File: packages/ra-core/package.json (hunk 7 above)
Comment (djhi) [reply packages/ra-core/package.json]: It's the default used by the `TranslationProvider`, which is provided by `ra-core`

File: packages/ra-core/src/ui/delete/Delete.js (hunk 10 above)
Comment (djhi) [reply packages/ra-core/src/ui/delete/Delete.js]: The `ra-ui-materialui` `Delete` component does

File: packages/ra-core/src/ui/detail/Show.js (hunk 14 above)
Comment (djhi) [reply packages/ra-core/src/ui/detail/Show.js]: This shouldn't be done by CoreShow at all actually. Moving it to the mui Show

File: packages/ra-core/src/ui/input/ReferenceArrayInput.js (hunk 19 above)
Comment (djhi) [reply packages/ra-core/src/ui/input/ReferenceArrayInput.js]: Because it was actually missing. A warning was displayed in the console when testing because it can become: `matchingReferences: { error: 'fetch error' }`

Checkout the unit tests

File: packages/ra-ui-materialui/package.json (hunk 23 above)
Comment (djhi) [reply packages/ra-ui-materialui/package.json]: Polyfill for the FileInput

File: packages/ra-ui-materialui/package.json (hunk 25 above)
Comment (djhi) [reply packages/ra-ui-materialui/package.json]: We don't use tilde for any references to react-admin packages in any of the other packages

File: packages/ra-core/package.json (hunk 7 above)
Comment (fzaninotto) [reply packages/ra-core/package.json]: :+1:

File: packages/ra-core/src/ui/input/ReferenceInput.js (hunk 21 above)
Comment (fzaninotto) [reply packages/ra-core/src/ui/input/ReferenceInput.js]: disregard, by bad

File: packages/ra-core/src/ui/input/ReferenceInput.js (hunk 20 above)
Comment (fzaninotto) [reply packages/ra-core/src/ui/input/ReferenceInput.js]: we agreed it shouldn't be the case

File: packages/ra-ui-materialui/package.json (hunk 25 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/package.json]: ok, we'll do that in another PR

File: Makefile [hunk 28]
```diff
@@ -74,7 +82,7 @@ doc: ## compile doc as html and launch doc web server
 
 lint: ## lint the code and check coding conventions
 	@echo "Running linter..."
-	@"./node_modules/.bin/eslint" ./packages/react-admin/src/
+	@"./node_modules/.bin/eslint" ./packages/react-admin/src/ ./packages/ra-core/src/
```
Comment (fzaninotto) [inline Makefile]: you should lint `ra-ui-materialui`, too

Comment (djhi) [reply Makefile]: In fact, we should lint all our projects ;)

Comment (djhi) [reply Makefile]: `./packages/**/src` ?
